### PR TITLE
attach cudagraph cache to the transform

### PR DESCRIPTION
**Pull request by t-vi**
This introduces a CUDAGraphRunner class to bundle building and caching cudagraphs and attach them to the CUDAGraphTransform introduced in #977 .

(Design issue #981 ) 

From my POV this is ready for review, but it builds on #977, so I'm labeling it draft.

### CODE REVIEW

**Comment by t-vi** [comment]
@nikitaved @IvanYashchuk

File: thunder/transforms/cudagraph.py [hunk 1]
```diff
@@ -100,44 +111,79 @@ def __call__(self, *args):
...
+            static_inputs_mask = None
+
+        x_fn_name = f"{fn_name}_{self.name_counter}"
+        self.name_counter += 1
+
+        self.python_callables[x_fn_name] = (
+            self.make_python_callable_from_symbols(fn_name, bsyms, inputs, outputs),
+            static_inputs_mask,
+        )
+        self.trace_symbols[x_fn_name] = (bsyms, inputs, outputs)
```
Comment (IvanYashchuk) [inline thunder/transforms/cudagraph.py:164]: How can I access these dictionaries in the regular user script? How should developers get to know that this information is saved and available for inspection?

File: thunder/transforms/cudagraph.py [hunk 2]
```diff
@@ -35,63 +34,75 @@ def extract_descriptor(arg):
...
-    # PyTorch's Inductor: torch/_inductor/cudagraph_trees.py
-    # Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/view
-    graph = torch.cuda.CUDAGraph()
-    with torch.cuda.graph(graph, stream=stream):
-        static_outputs = fn(*static_inputs)
-
-    return graph, static_inputs, static_outputs
+    def build_cuda_graph(
+        self, fn: Callable, args: list[any], static_args_mask: tuple[bool, ...]
+    ) -> tuple[torch.cuda.CUDAGraph, Sequence[torch.Tensor | Any], Sequence[torch.Tensor | Any]]:
```
Comment (IvanYashchuk) [inline thunder/transforms/cudagraph.py:54]: What is the benefit of having this as a method and not as a free function?

**Review by IvanYashchuk — APPROVED** [review]
The code is organized a bit differently now. Looking forward to where it will take us!

File: thunder/transforms/cudagraph.py [hunk 3]
```diff
@@ -34,105 +34,155 @@ def extract_descriptor(arg):
...
-        self.num_static_inputs = num_static_inputs
+        graph.replay()
+        return static_outputs
 
-    def __call__(self, *args):
-        if self.num_static_inputs is not None:
-            static_inputs_mask = (True,) * self.num_static_inputs + (False,) * (len(args) - self.num_static_inputs)
-        else:
-            static_inputs_mask = tuple(isinstance(arg, torch.nn.Parameter) for arg in args)
+    def make_python_callable_from_symbols(
```
Comment (lantiga) [inline thunder/transforms/cudagraph.py:114]: This looks like something we could have in a transform building block library in the future

File: thunder/transforms/cudagraph.py (hunk 1 above)
Comment (t-vi) [reply thunder/transforms/cudagraph.py:164]: Currently, you have to look at the source. I'm not yet sure that this is the exact information we want to keep, we should refine this.

File: thunder/transforms/cudagraph.py (hunk 2 above)
Comment (t-vi) [reply thunder/transforms/cudagraph.py:54]: My tinking was that having the ability to override the buffer allocation (get_static_buffer) could get useful. We could have it as a callback instead but then it might just as well be a method.

**Review by lantiga — APPROVED** [review]
Looks great

File: thunder/transforms/cudagraph.py (hunk 3 above)
Comment (t-vi) [reply thunder/transforms/cudagraph.py:114]: Indeed! The question is whether we should build the region before...

File: thunder/transforms/cudagraph.py [hunk 4]
```diff
@@ -100,44 +111,79 @@ def __call__(self, *args):
...
+
+        return callable
 
 
 class CUDAGraphTransform(Transform):
     """
     Transform to fuse operations into CUDA graphs post optimization.
 
-    This class provides the basic infrastructure, but it is expected that you might subclass this transform
+    This class provides the basic infrastructure, but it is expected that you might subclass this transformm
```
[the remark this answers is not in the document]
Comment (t-vi) [reply thunder/transforms/cudagraph.py]: fixed, thank you!

File: thunder/transforms/cudagraph.py (hunk 1 above)
Comment (t-vi) [reply thunder/transforms/cudagraph.py:164]: so in addition to everything else the `x_fn_name` is a bit stupid, the trick was that the transform is applied to forward and backward. If we are OK with cudagraph regions not starting from 1 in the backward, I would just generate the name in the runner and return it.
This would allow a link from trace->cache name here easily. Is that what you have in mind?

**Comment by t-vi** [comment]
So I'm merging this now, I expect to do another PR or two in the next few days around the dev x of this.
(One other thing I want to do is perhaps pass in the static argument mask which we might obtain from analysing the graph(s) - parameters+buffers from the module are static etc.)
